src/ucp/am/eager.inl

```diff
@@ -63,4 +66,68 @@ ucp_am_bcopy_pack_data(void *buffer, ucp_request_t *req, size_t length,
     return total_length;
 }
 
+static void ucp_proto_request_eager_am_zcopy_completion(uct_completion_t *self)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
+
+    ucs_assert(req->send.msg_proto.am.reg_desc != NULL);
+    ucs_mpool_put_inline(req->send.msg_proto.am.reg_desc);
+    ucp_proto_request_zcopy_completion(self);
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_eager_am_zcopy_pack_user_header(ucp_request_t *req)
+{
+    ucp_mem_desc_t *reg_desc;
+
+    reg_desc = ucp_worker_mpool_get(&req->send.ep->worker->reg_mp);
+    if (ucs_unlikely(reg_desc == NULL)) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (req->send.msg_proto.am.header_length != 0) {
+        ucs_assert(req->send.msg_proto.am.header != NULL);
+        ucp_am_pack_user_header(reg_desc + 1, req);
+    }
+
+    req->send.msg_proto.am.reg_desc = reg_desc;
```

Comment:
maybe skip allocating desc if header_length==0, like it was before?

Reply:
we also use it for footers

Reply:
We can only skip the allocation if header_length==0 in case of single fragment zcopy.
We use this desc to send footer in case of multi-fragment zcopy,

Reply:
but only for multi-fragment footer AFAIU?

Reply:
Do you think it would be better to use to init methods for single and multi zcopy instead of having a common function?
E.g. introduce variable to pass to `ucp_eager_am_zcopy_pack_user_header` through `ucp_proto_request_eager_am_zcopy_init`.

Reply:
i'd go with always getting an rdesc, like it is done in old protocols